### PR TITLE
Certificatemanager

### PR DESCRIPTION
**Pull request by ajaykangare**
### Description of your changes

Create a Certificate as a managed resource.

Example: example/certificate/

### Checklist

I have:
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Ensured this PR contains a neat, self documenting set of commits.
- [ ] Updated any relevant [documentation], [examples], or [release notes].
- [ ] Updated the dependencies in [`app.yaml`] to include any new role permissions.

[documentation]: https://github.com/crossplane/crossplane/tree/master/docs
[examples]: https://github.com/crossplane/crossplane/tree/master/cluster/examples
[release notes]: https://github.com/crossplane/crossplane/tree/master/PendingReleaseNotes.md
[`app.yaml`]: https://github.com/crossplane/provider-aws/blob/master/config/stack/manifests/app.yaml

### CODE REVIEW

File: pkg/clients/certificatemanager/certificate.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+		m.Options = &acm.CertificateOptions{CertificateTransparencyLoggingPreference: acm.CertificateTransparencyLoggingPreferenceDisabled}
+	} else if strings.EqualFold(p.CertificateTransparencyLoggingPreference, "ENABLED") {
+		m.Options = &acm.CertificateOptions{CertificateTransparencyLoggingPreference: acm.CertificateTransparencyLoggingPreferenceEnabled}
+	}
+
+	if strings.EqualFold(p.ValidationMethod, "EMAIL") {
+		m.ValidationMethod = acm.ValidationMethodEmail
+	} else if strings.EqualFold(p.ValidationMethod, "DNS") {
+		m.ValidationMethod = acm.ValidationMethodDns
+	}
```
Comment (hasheddan) [inline pkg/clients/certificatemanager/certificate.go:51]: Would be good to make these `enum` types on the CRD, ref: https://github.com/crossplane/provider-aws/blob/d2ddf8ff9521dcbfbeba2215f35f9b401414e1ef/apis/network/v1alpha3/internetgateway_types.go#L52

File: pkg/clients/certificatemanager/certificate.go [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+	if len(p.Tags) != len(tags) {
+		return false
+	}
+
+	pTags := make(map[string]string, len(p.Tags))
+	for _, tag := range p.Tags {
+		pTags[tag.Key] = tag.Value
+	}
```
Comment (hasheddan) [inline pkg/clients/certificatemanager/certificate.go:156]: Does `p.Tags` need to be copied into a new map for comparison?

File: apis/certificatemanager/v1alpha1/certificate_types.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	runtimev1alpha1.ResourceStatus `json:",inline"`
+	AtProvider                     CertificateExternalStatus `json:"atProvider"`
+}
+
+// CertificateParameters defines the desired state of an AWS Certificate.
+type CertificateParameters struct {
+
+	// The Amazon Resource Name (ARN) of the private certificate authority (CA)that will be used to issue the certificate.
+	// +optional
+	CertificateAuthorityArn *string `json:"certificateAuthorityArn,omitempty"`
```
[the remark this answers is not in the document]
Comment (ajaykangare) [reply apis/certificatemanager/v1alpha1/certificate_types.go:72]: Added the _Ref and _Selector fields but as it needs a reference of CertificateAuthority so commented the code for the same. OnceCertificateAuthority PR is merged, we can do the needful.

File: apis/certificatemanager/v1alpha1/certificate_types.go [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+	// +optional
+	DomainValidationOptions []*DomainValidationOption `json:"domainValidationOptions,omitempty"`
+
+	// Token to distinguish between calls to RequestCertificate.
+	// +optional
+	IdempotencyToken *string `json:"idempotencyToken,omitempty"`
+
+	// Parameter add the certificate to a certificate transparency log.
+	// +optional
+	CertificateTransparencyLoggingPreference string `json:"certificateTransparencyLoggingPreference,omitempty"`
```
[the remark this answers is not in the document]
Comment (ajaykangare) [reply apis/certificatemanager/v1alpha1/certificate_types.go]: Changed the type to Enum.

File: apis/certificatemanager/v1alpha1/certificate_types.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+	// +optional
+	SubjectAlternativeNames []string `json:"subjectAlternativeNames,omitempty"`
+
+	// One or more resource tags to associate with the certificate.
+	// +optional
+	Tags []Tag `json:"tags,omitempty"`
+
+	// Method to validate certificate.
+	// +optional
+	ValidationMethod string `json:"validationMethod,omitempty"`
```
[the remark this answers is not in the document]
Comment (ajaykangare) [reply apis/certificatemanager/v1alpha1/certificate_types.go]: Changed the type to Enum

File: apis/certificatemanager/v1alpha1/certificate_types.go [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+
+	// Flag to renew the certificate
+	// +optional
+	RenewCertificate bool `json:"renewCertificate,omitempty"`
```
[the remark this answers is not in the document]
Comment (ajaykangare) [reply apis/certificatemanager/v1alpha1/certificate_types.go]: Added pointer

File: apis/certificatemanager/v1alpha1/v1alpha1_suite_test.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/envtest"
+)
+
+var cfg *rest.Config
+var c client.Client
+
+func TestMain(m *testing.M) {
```
[the remark this answers is not in the document]
Comment (ajaykangare) [reply apis/certificatemanager/v1alpha1/v1alpha1_suite_test.go]: Removed file

File: pkg/clients/certificatemanager/certificate.go (hunk 1 above)
Comment (ajaykangare) [reply pkg/clients/certificatemanager/certificate.go]: Thanks, Changed the type to Enum.

File: pkg/clients/certificatemanager/certificate/certificate_test.go [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+	domainName                 = "somedomain"
+	certificateArn             = "somearn"
+	renewalEligibilityEligible = "ELIGIBLE"
+	certificateAuthorityArn    = "someauthorityarn"
+)
+
+// func addCertificateOutputFields(c *acm.CertificateDetail) {
+// 	c.CertificateArn = aws.String(certificateArn)
+// 	c.RenewalEligibility = acm.RenewalEligibilityEligible
+// }
```
[the remark this answers is not in the document]
Comment (ajaykangare) [reply pkg/clients/certificatemanager/certificate/certificate_test.go:23]: Removed commented code

**Pull request closed without merging**